vm/systemSmartContracts/auction.go

```diff
@@ -0,0 +1,751 @@
+}
+
+// AuctionConfig represents the settings for a specific epoch
+type AuctionConfig struct {
+	MinStakeValue *big.Int `json:"MinStakeValue"`
+	NumNodes      uint32   `json:"NumNodes"`
+	TotalSupply   *big.Int `json:"TotalSupply"`
+	MinStep       *big.Int `json:"MinStep"`
+	NodePrice     *big.Int `json:"NodePrice"`
+}
+
+type stakingAuctionSC struct {
+	eei           vm.SystemEI
+	unBoundPeriod uint64
+	kg            crypto.KeyGenerator
+	baseConfig    AuctionConfig
+}
+
+// NewStakingAuctionSmartContract creates an auction smart contract
+func NewStakingAuctionSmartContract(
+	minStakeValue *big.Int,
+	minStepValue *big.Int,
+	totalSupply *big.Int,
+	unBoundPeriod uint64,
+	numNodes uint32,
+	eei vm.SystemEI,
+	kg crypto.KeyGenerator,
+) (*stakingAuctionSC, error) {
+	if minStakeValue == nil {
+		return nil, vm.ErrNilInitialStakeValue
+	}
+	if minStakeValue.Cmp(big.NewInt(0)) < 1 {
+		return nil, vm.ErrNegativeInitialStakeValue
+	}
+	if check.IfNil(eei) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+
+	baseConfig := AuctionConfig{
+		MinStakeValue: minStakeValue,
+		NumNodes:      numNodes,
+		TotalSupply:   totalSupply,
+		MinStep:       minStepValue,
+		NodePrice:     big.NewInt(0).Set(minStakeValue),
+	}
+
+	reg := &stakingAuctionSC{
+		eei:           eei,
+		unBoundPeriod: unBoundPeriod,
+		kg:            kg,
+		baseConfig:    baseConfig,
+	}
+	return reg, nil
+}
+
+// Execute calls one of the functions from the staking smart contract and runs the code according to the input
+func (s *stakingAuctionSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case "_init":
+		return s.init(args)
+	case "stake":
+		return s.stake(args)
+	case "unStake":
+		return s.unStake(args)
+	case "unBound":
+		return s.unBound(args)
+	case "claim":
+		return s.claim(args)
+	case "slash":
+		return s.slash(args)
+	case "get":
+		return s.get(args)
+	case "setConfig":
+		return s.setConfig(args)
+	}
+
+	//TODO: integrated into the protocol the calling the functions select and setConfig at end-of-epoch
+
+	return vmcommon.UserError
+}
+
+func (s *stakingAuctionSC) get(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) < 1 {
+		return vmcommon.UserError
+	}
+
+	value := s.eei.GetStorage(args.Arguments[0])
+	s.eei.Finish(value)
+
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) setConfig(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	ownerAddress := s.eei.GetStorage([]byte(ownerKey))
+	if !bytes.Equal(ownerAddress, args.CallerAddr) {
+		log.Debug("setConfig function was not called by the owner address")
+		return vmcommon.UserError
+	}
+
+	if len(args.Arguments) != 6 {
+		log.Debug("setConfig function called with wrong number of arguments")
+		return vmcommon.UserError
+	}
+
+	config := AuctionConfig{
+		MinStakeValue: big.NewInt(0).SetBytes(args.Arguments[0]),
```

Comment:
this will be set by the protocol at end-of-epoch. checks added.

Reply:
ok

Reply:
Will be added when integration is done. It will be checked in every part, protocol, endofepoch and contract as well. Arguments may change when integrated

Reply:
ok